source/common/http/http1/codec_impl.cc

```diff
@@ -442,9 +442,18 @@ ConnectionImpl::ConnectionImpl(Network::Connection& connection, Stats::Scope& st
       reject_unsupported_transfer_encodings_(Runtime::runtimeFeatureEnabled(
           "envoy.reloadable_features.reject_unsupported_transfer_encodings")),
       output_buffer_([&]() -> void { this->onBelowLowWatermark(); },
-                     [&]() -> void { this->onAboveHighWatermark(); }),
+                     [&]() -> void { this->onAboveHighWatermark(); },
+                     [&]() -> void { this->onAboveOverflowWatermark(); }),
       max_headers_kb_(max_headers_kb), max_headers_count_(max_headers_count) {
-  output_buffer_.setWatermarks(connection.bufferLimit());
+
+  const uint32_t connection_buffer_limit = connection.bufferLimit();
+  if (connection_buffer_limit == 0) {
+    // Disable watermarking
+    output_buffer_.setWatermarks(0);
+  } else {
+    output_buffer_.setWatermarks(connection_buffer_limit / 2, connection_buffer_limit,
+                                 connection_buffer_limit + 1);
```

Comment:
I think you are correct, and according to: https://github.com/envoyproxy/envoy/blob/cfa29e187007331a93bfc83df5873aa7270b3980/source/common/network/connection_impl.cc#L451-L455

Both this and the H2's codec should be 2*HW.
Regarding configurability, it could be added next to every HW configuration parameter.
IMO it improves the usability of OW, and slightly increases configuration complexity. It makes sense to add this.

Reply:
Roughly 2x data is the common case, so I agree that you need at least that much.  There are examples of H1 to H2 conversion that can result in the limit being exceeded further beyond the 2x HW.  Another example is serialization of headers together with parts of the body, where you can end up with serialized headers + HW worth of body in the output buffer.  We do not require that the header size limit is below the HW.  In the case of H2, you can end up with control frames in the output buffer in addition to 2x HW worth of HTTP message bodies.

We have also talked about relationship between H2 remote endpoint connection window and how it relates to max size of the output buffer.  It would be good to understand how big output buffers can get beyond HW before switching over to actions beyond incrementing stat counters.